keps/sig-apps/0008-sidecarcontainer.md

```diff
@@ -0,0 +1,116 @@
+
+The only way around this problem is to manage the sidecar container's lifecycle manually and arrange for it to exit when the main container exits. This is typically achieved by building an ad-hoc signalling mechanism to communicate completion status between containers. Common implementations use a shared scratch volume mounted into all pods, where lifecycle status can be communicated by creating and watching for the presence of files. This pattern has several disadvantages:
+
+* Repetitive lifecycle logic must be rewritten in each instance a sidecar is deployed.
+* Third-party containers typically require a wrapper to add this behaviour, normally provided via an entrypoint wrapper script implemented in the k8s container spec. This adds undesirable overhead and introduces repetition between the k8s and upstream container image specs.
+* The wrapping typically requires the presence of a shell in the container image, so this pattern does not work for minimal containers which ship without a toolchain.
+
+### Goals
+
+Solve issue [25908](https://github.com/kubernetes/kubernetes/issues/25908) so that sidecars can be easily handled without needing modifications to the container itself
+
+### Non-Goals
+
+Changing the way jobs are handled
+
+## Proposal
+
+Create a way to define containers as sidecars, this will be an additional field to the Container Spec: `sidecar: true`.
+e.g:
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: myapp-pod
+  labels:
+    app: myapp
+spec:
+  containers:
+  - name: myapp
+    image: myapp
+    command: ['do something']
+  - name: sidecar
+    image: sidecar-image
+    sidecar: true
+    command: ["do something to help my app"]
+
+```
+These sidecar containers will be handled the same way as a non-sidecar container in almost all respects apart from the fact that when all non-sidecar containers have permanently terminated all running sidecar containers will also be terminated with a reason of `Completed`. The sidecars will be terminated by receiving a SIGTERM and will obey the `terminationGracePeriodSeconds` defined on the pod before being sent a SIGKILL. Sidecar containers will be terminated in the same way as non-sidecars when a pod is deleted.
+
+For example, in a job you can have as many sidecars as you like and all the running sidecars will be terminated when all the non-sidecar containers have terminated, causing your job to finish.
+If all the non-sidecars `Completed` the pod status will be `Succeeded`, if any of the non-sidecars terminated with `Error` the pod status will be `Failed` and if any of the sidecars terminated with `Error` before the non-sidecars terminated, the pod status will be `Failed`.
```

Comment:
How would this interact with restart policies `Always` and `OnFailure`?

IFor the use-cases listed in issue [52345](https://github.com/kubernetes/kubernetes/issues/52345) and [this draft KEP](https://github.com/kubernetes/community/pull/2342), can this feature be used in Deployments or ReplicaSets with `restartPolicy` `Always`?

I.e., with `restartPolicy` `Always`, if all the non-sidecar containers terminate (with success or error), would the sidecar containers be terminated and the pod marked as `completed` (with status `Succeeded` or `Error`)? If yes, I will close the draft KEP in favour of this proposal.

Reply:
@amshuman-kr so if your restart policy is 'Always' all containers will always restart on failure/success which is no different to how it currently works. If it's "OnFailure" your non-sidecars will only restart if they Error, sidecars will do the same, once all non-sidecars exit with success the sidecars will all be terminated.

Reply:
@Joseph-Irving Thanks for the clarification! My question was based on the requirement to re-execute init-containers when the non-sidecar containers terminate (with success or error).

> if your restart policy is `Always` all containers will always restart on failure/success

With restart policy `Always`, if a non-sidecar container terminates, will the init-containers also be re-executed? I.e. will the whole `Pod` be restarted? Or will only the terminated non-sidecar container be restarted?

From your explanation, it seems to be the latter. But then this would limit the extent of the linking of life-cycle of a Pod to the life-cycle of non-sidecar containers. It would not extend well to ReplicaSets etc. and won't address [52345](https://github.com/kubernetes/kubernetes/issues/52345). Is my reading right?

Reply:
@amshuman-kr, yes only the container will restart not the init containers, which is what normally happens, so you want a non-sidecar restart to cause an init container to re-run? If so I don't think that's in scope for this proposal.

Reply:
@Joseph-Irving Understood. @thockin and @erictune pointed me to the idea in [this](https://github.com/kubernetes/community/pull/2342#pullrequestreview-134882595) [proposal](https://github.com/kubernetes/community/pull/2342#issuecomment-403546905) as a better alternative to my [draft](https://github.com/kubernetes/community/pull/2342) proposal. Is there any chance of taking this requirement also onboard in this proposal? I.e. the requirement about re-executing init-containers on non-sidecar restart?